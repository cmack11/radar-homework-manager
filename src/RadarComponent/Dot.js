import React, { Component } from 'react';
//import moment from 'moment'
//import PropTypes from 'prop-types';

class Dot extends Component {


	constructor(props) {
		super();
		this.minOpacity = .1;
		this.opacityStep = .03;
		this.state = {opacity:this.minOpacity}
		
	}

	startFadeOut() {
		this.fadeOut();
		this.fOut = setInterval(() => {
			this.fadeOut();
			if(this.state.opacity <= this.minOpacity) {
				clearInterval(this.fOut);
				this.fOut = null;
			}

		},400)
 	}

	fadeOut() {
		let opacity = Math.max(this.minOpacity,this.state.opacity - this.opacityStep);
		this.setState({opacity:opacity});
	}

	componentDidMount() {
		if(this.props.dot)
			this.props.dot.restartFadeOut = this.restartFadeOut.bind(this);
		if(this.props.animateFades) {
			this.checkIntersect = setInterval(() => {
				if(this.props.intersectsLine(this.props.dot,this.getId())){
					this.startFadeIn();
				}
			},50)
		} else {
			this.setState({opacity:1})
		}
	}

	componentWillUnmount() {
		if(this.checkIntersect)
			clearInterval(this.checkIntersect);
		if(this.fOut)
			clearInterval(this.fOut);
	}

	getId() {
		return this.props.center.x+'/'+this.props.center.y;
	}

	startFadeIn() {
		if(this.fOut) {
			clearInterval(this.fOut);
			this.fOut = null;
		}
		this.fadeIn();
		this.startFadeOut();
		/*let fIn = setInterval(() => {
			this.fadeIn();
			if(this.state.opacity >= 1) {
				this.fadeOut();
				clearInterval(fIn);
			}
		},300)*/
	}

	fadeIn() {
			let opacity = 1//Math.min(1,this.state.opacity + (this.opacityStep*50));
			this.setState({opacity:opacity});
	}

	restartFadeOut() {
		this.fadeOut();
	}

	onMouseDown(e) {
		this.props.onMouseDown(e,this.props.dot);
	}


	render() {
		if(!this.props.center || this.props.center.x < 0 || this.props.center.y < 0 || this.props.radius < 0)
			return null;

		return (
			<circle id={this.getId()} cx={this.props.center.x} cy={this.props.center.y} r={this.props.radius} fill={this.props.fill}
				 draggable={true} opacity={this.state.opacity} 
				 onMouseDown={this.onMouseDown.bind(this)}
				 cursor='pointer'
				 />

		)
	}
}

export default Dot;

//need to find a way to have the dragged dot be on top of the spinline, but stationary dots be below
//And also have dotviewer be on top of spinline

