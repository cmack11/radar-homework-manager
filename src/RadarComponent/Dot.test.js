import React from 'react';
import Dot from './Dot.js';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import * as Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from "sinon";

beforeAll(() => {
  Enzyme.configure({ adapter: new Adapter() })
  jest.useFakeTimers();
});

const dot = {
  center:{x:10,y:15},
  radius:100,
}

describe('the initialization of the component', () => {

  it('creates an svg circle with the correct attributes', () => {
    const wrapper = shallow(
      <Dot center={dot.center} radius={dot.radius} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('creates a null component because of invalid x prop', () => {
    const wrapper = shallow(
      <Dot center={{x:-50, y:dot.center.y}} radius={dot.radius} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('creates a null component because of invalid y prop', () => {
    const wrapper = shallow(
      <Dot center={{x:dot.center.x, y:-100}} radius={dot.radius} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('creates a null component because of invalid radius prop', () => {
    const wrapper = shallow(
      <Dot center={dot.center} radius={-1} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('initialize the dot to have the minimum opacity', () => {
    const wrapper = shallow(
      <Dot center={dot.center} radius={dot.radius} animateFades={true}/>,
    );

    expect(wrapper).toMatchSnapshot();
  });

})


describe('the fadeIn function', () => {

  it('should fade in by setting the opacity instantly to 1', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} />);
    component.instance().fadeIn()

    expect(component.instance().state.opacity).toEqual(1);
  });

})

describe('the checkIntersect function', () => {

  it('should fill checkIntersect with a reference to the asych function that checks for the spinline to intersect this function', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} animateFades={true} />);
    
    expect(component.instance().checkIntersect).toBeDefined();
  });

});

describe('the restartFadeOut function', () => {

  it('should called fadeOut when restartFadeOut is called', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} animateFades={true} />);
    const spy = jest.spyOn(component.instance(),'fadeOut');
    
    component.instance().restartFadeOut();
    
    expect(spy).toBeCalled();
  });

})


describe('the startFadeIn function', () => {

  it('should clear the old fadeOut interval', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} animateFades={true} />);
    component.instance().fOut = {};

    component.instance().startFadeIn();
    
    expect(clearInterval).toBeCalled();
  });

  it('should call the fadeIn function and the startFadeOut function', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} animateFades={true} />);
    const spy1 = jest.spyOn(component.instance(),'fadeIn');
    const spy2 = jest.spyOn(component.instance(),'startFadeOut');

    component.instance().startFadeIn();
    
    expect(spy1).toBeCalled();
    expect(spy2).toBeCalled();
  });

  it('should call fadeIn and then call startFadeOut', () => {
    const component = shallow(<Dot center={dot.center} radius={dot.radius} animateFades={true} />);
    const spy1 = jest.spyOn(component.instance(),'fadeIn');
    const spy2 = jest.spyOn(component.instance(),'startFadeOut');

    component.instance().startFadeIn();
    
    expect(spy1).toBeCalled();
    expect(spy2).toBeCalled();
  });

})

