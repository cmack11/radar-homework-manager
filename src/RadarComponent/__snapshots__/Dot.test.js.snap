// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the initialization of the component creates a null component because of invalid radius prop 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    radius={-1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`the initialization of the component creates a null component because of invalid x prop 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    center={
      Object {
        "x": -50,
        "y": 15,
      }
    }
    radius={100}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`the initialization of the component creates a null component because of invalid y prop 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    center={
      Object {
        "x": 10,
        "y": -100,
      }
    }
    radius={100}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`the initialization of the component creates an svg circle with the correct attributes 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    dot={
      Object {
        "restartFadeOut": [Function],
      }
    }
    radius={100}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "cursor": "default",
      "cx": 10,
      "cy": 15,
      "draggable": true,
      "fill": undefined,
      "id": "10/15",
      "onMouseDown": [Function],
      "opacity": 1,
      "r": 100,
    },
    "ref": null,
    "rendered": null,
    "type": "circle",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "cursor": "default",
        "cx": 10,
        "cy": 15,
        "draggable": true,
        "fill": undefined,
        "id": "10/15",
        "onMouseDown": [Function],
        "opacity": 1,
        "r": 100,
      },
      "ref": null,
      "rendered": null,
      "type": "circle",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`the initialization of the component initialize the dot to have the minimum opacity 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    animateFades={true}
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    intersectsLine={[MockFunction]}
    radius={100}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "cursor": "default",
      "cx": 10,
      "cy": 15,
      "draggable": true,
      "fill": undefined,
      "id": "10/15",
      "onMouseDown": [Function],
      "opacity": 0.1,
      "r": 100,
    },
    "ref": null,
    "rendered": null,
    "type": "circle",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "cursor": "default",
        "cx": 10,
        "cy": 15,
        "draggable": true,
        "fill": undefined,
        "id": "10/15",
        "onMouseDown": [Function],
        "opacity": 0.1,
        "r": 100,
      },
      "ref": null,
      "rendered": null,
      "type": "circle",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`the onMouseDown function should call the prop onMouseDown function 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dot
    animateFades={false}
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    dot={
      Object {
        "restartFadeOut": [Function],
      }
    }
    intersectsLine={[MockFunction]}
    onMouseDown={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
            Object {
              "restartFadeOut": [Function],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    radius={100}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "cursor": "default",
      "cx": 10,
      "cy": 15,
      "draggable": true,
      "fill": undefined,
      "id": "10/15",
      "onMouseDown": [Function],
      "opacity": 1,
      "r": 100,
    },
    "ref": null,
    "rendered": null,
    "type": "circle",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "cursor": "default",
        "cx": 10,
        "cy": 15,
        "draggable": true,
        "fill": undefined,
        "id": "10/15",
        "onMouseDown": [Function],
        "opacity": 1,
        "r": 100,
      },
      "ref": null,
      "rendered": null,
      "type": "circle",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
